File: __LEGACY__/al-leetcode/uncategorized/dp_coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # You are given an integer array coins representing coins of different denominations and an integer amount representing a total amount of money.
        # Return the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.
        # You may assume that you have an infinite number of each kind of coin.

        # Dynamic Programming
        # Time: O(n*amount)
        # Space: O(amount)
        if not coins:
            return -1
        if amount == 0:
            return 0
        dp = [0 for _ in range(amount + 1)]
        dp[0] = 0
        for i in range(1, amount + 1):
            dp[i] = float('inf')
            for coin in coins:
                if i - coin >= 0 and dp[i - coin] != float('inf'):
                    dp[i] = min(dp[i], dp[i - coin] + 1)
        return dp[-1] if dp[-1] != float('inf') else -1

        

File: __LEGACY__/al-leetcode/uncategorized/dp_coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    // You are given coins of different denominations and a total amount of money amount. Write a function to compute the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.
    // Example 1:
    // Input: coins = [1, 2, 5], amount = 11
    // Output: 3 
    // Explanation: 11 = 5 + 5 + 1
    // Example 2:
    // Input: coins = [2], amount = 3
    // Output: -1
    // Note:
    // You may assume that you have an infinite number of each kind of coin.
    // 
    // 
    // Dynamic Programming
    // O(n * amount)
    if (amount === 0) {
        return 0;
    }
    if (coins.length === 0) {
        return -1;
    }
    let dp = new Array(amount + 1).fill(0);
    dp[0] = 0;
    for (let i = 1; i <= amount; i++) {
        for (let j = 0; j < coins.length; j++) {
            if (i - coins[j] >= 0) {
                dp[i] = Math.min(dp[i], dp[i - coins[j]] + 1);
            }
        }
    }
    return dp[amount] === 0 ? -1 : dp[amount];    
};